### PR TITLE
added valley functions

### LINKED ISSUE #5: Add Valley-Shaped functions
Add missing functions from Valley-Shaped category from [SFU](https://www.sfu.ca/~ssurjano/optimization.html)

### PR DESCRIPTION
**Pull request by dillonago**
Please let me know if it works. Thanks!

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,63 @@
+{
```
Comment (kvoss) [inline .vscode/settings.json]: let's remove this file from the commit

File: bss.cc [hunk 2]
```diff
@@ -555,6 +555,35 @@ trefethen4(const std::vector<double>& xs)
     return ret;
 }
 
+//Valley Functions
+double
+threeHumpCamel(const std::vector<double>& xs)
```
Comment (kvoss) [inline bss.cc]: let's stick to only lower case and underscores between words in naming

File: bss.cc [hunk 3]
```diff
@@ -555,6 +555,35 @@ trefethen4(const std::vector<double>& xs)
...
+dixonPrice(const std::vector<double>& xs)
+{
+    double x1 = xs.at(0);
+    double acc0 = 0;
+    for(std::size_t d=2; d <= xs.size(); ++d) {
+        acc0 += d*SQ((2 * SQ(xs.at(d-1))-xs.at(d-2)));
+    }
+
+    double ret = SQ(x1-1) 
+    + acc0;
```
Comment (kvoss) [inline bss.cc]: this might be better on the previous line

File: bss.cc [hunk 4]
```diff
@@ -555,6 +555,35 @@ trefethen4(const std::vector<double>& xs)
...
 }
 
+//Valley Functions
+double
+threeHumpCamel(const std::vector<double>& xs)
+{
+    double x1 = xs.at(0);
+    double x2 = xs.at(1);
+
+    double ret = (2 * SQ(x1))
```
Comment (kvoss) [inline bss.cc:565]: for the style consistency, I'd recommend using " \" at the end of a line

File: cbench.cc [hunk 5]
```diff
@@ -71,6 +71,8 @@ PYWRAP(shubert);
 PYWRAP(sphere);
 PYWRAP(tripod);
 PYWRAP(trefethen4);
+PYWRAP(threeHumpCamel);
```
Comment (kvoss) [inline cbench.cc]: only naming in this file

File: bss.h [hunk 6]
```diff
@@ -37,6 +37,8 @@ double sphere(const std::vector<double>& xs);
 double parabola(const std::vector<double>& xs);
 double tripod(const std::vector<double>& xs);
 double trefethen4(const std::vector<double>& xs);
+double threeHumpCamel(const std::vector<double>& xs);
```
Comment (kvoss) [inline bss.h]: three_hump_camel_back

**Review by kvoss — CHANGES_REQUESTED** [review]
nicely done! There is only a few small things

**Comment by kvoss** [comment]
@dillonago one more thing, after changes, run `make && make test`. That should show you differences with the "golden" file. That's my current safety net. Just add the differences to the golden and add that as part of a commit.

File: bss.cc (hunk 4 above)
Comment (dillonago) [reply bss.cc:565]: Sorry, I don't understand this. What do you mean by adding " "?

File: bss.cc (hunk 3 above)
Comment (dillonago) [reply bss.cc]: What might be better? Sorry, I don't see any changes.

**Comment by dillonago** [comment]
I still need clarification on the suggestions, and have to get my env up and running to get the differences.

**Comment by dillonago** [comment]
Hi @kvoss , I added my outputs to the bottom of the "golden" file. Please let me know what else needs to be tweaked before this can get merged. I will get on it right away. Thanks!

File: bss.cc (hunk 3 above)
Comment (kvoss) [reply bss.cc]: it's cosmetic: just to keep it in one line: `double ret = SQ(x1-1)  + acc0`

File: bss.cc (hunk 4 above)
Comment (kvoss) [reply bss.cc:565]: this is my bad. I thought about adding " \" as in <space><backslash>.

Comment (dillonago) [reply bss.cc:565]: Ah gotchu you're good!

**Review by kvoss — CHANGES_REQUESTED** [review]
currently the `./testit.sh` script runs all functions and dumps them to `sanity-test-instance.log` overwriting the previous content of that file. Next, we compare to `sanity-test-golden.log`. The repo should have only the golden file, and instance is meant to be local.

Sorry about the confusing logic.

Could you remove the binary files and the `sanity-test-instance.log`?

One extra thing.. I try to add the names of the function names also to the list in the `README.md`.

That should be all :)

**Review by kvoss — CHANGES_REQUESTED** [review]
there should be only two extra lines in the `sanity-test-golden.log` file.

try removing the `sanity-test-instance.log`
then run `make test`
and that should show you where to add the two lines and what they are

**Comment by dillonago** [comment]
Hi again @kvoss, I think I did everything you asked. If not, let me know. Thanks for everything again!

File: bss.cc [hunk 7]
```diff
@@ -555,6 +555,34 @@ trefethen4(const std::vector<double>& xs)
...
+    double ret = (2 * SQ(x1)) 
+        - (1.05  * std::pow(x1, 4))
+        + (std::pow(x1, 6) / 6)
+        + (x1 * x2)
+        + SQ(x2);
+    return ret;
+}
+
+double 
+dixonPrice(const std::vector<double>& xs)
```
Comment (kvoss) [inline bss.cc]: this function name still mixes cases, `dixon_price` would be consistent with the rest

File: bss.h [hunk 8]
```diff
@@ -37,6 +37,8 @@ double sphere(const std::vector<double>& xs);
 double parabola(const std::vector<double>& xs);
 double tripod(const std::vector<double>& xs);
 double trefethen4(const std::vector<double>& xs);
+double three_hump_camel_back(const std::vector<double>& xs);
+double dixonPrice(const std::vector<double>& xs);
```
Comment (kvoss) [inline bss.h]: also here

File: cbench.cc [hunk 9]
```diff
@@ -71,6 +71,8 @@ PYWRAP(shubert);
 PYWRAP(sphere);
 PYWRAP(tripod);
 PYWRAP(trefethen4);
+PYWRAP(three_hump_camel_back);
+PYWRAP(dixonPrice);
```
Comment (kvoss) [inline cbench.cc]: `dixon_price`

File: cbench.cc [hunk 10]
```diff
@@ -109,6 +111,8 @@ CBenchMethods[] = {
     {"sphere", py_sphere, METH_VARARGS, "sphere function"},
     {"tripod", py_tripod, METH_VARARGS, "tripod function"},
     {"trefethen4", py_trefethen4, METH_VARARGS, "trefethen4 function"},
+    {"three_hump_camel_back", py_three_hump_camel_back, METH_VARARGS, "three hump camel function"},
+    {"dixonPrice", py_dixonPrice, METH_VARARGS, "dixon price function"},
```
Comment (kvoss) [inline cbench.cc]: and `dixon_price` and `py_dixon_price`

File: sanity-test-golden.log [hunk 11]
```diff
@@ -5,6 +5,7 @@
 -> cross_in_tray: -1.4777274811045142
 -> deceptive3: -0.0009897909414337454
 -> dejong5: 12.670567474629571
+-> dixonPrice: 0.8858402000000001
```
Comment (kvoss) [inline sanity-test-golden.log]: This is nearly perfect!!
that's exactly what I meant 👍

File: sanity-test-golden.log [hunk 12]
```diff
@@ -30,5 +31,11 @@
 -> six_hump_camel_back: -0.09380966666666667
 -> sphere: 0.5601
 -> step: 1.25
+-> three_hump_camel_back: 0.07989516666666668
 -> trefethen4: -0.13771083406418394
 -> tripod: 101.69999999999999
+
```
Comment (kvoss) [inline sanity-test-golden.log:37]: we can remove these empty lines at the end

**Review by kvoss — CHANGES_REQUESTED** [review]
a small adjustment to keep the names consistent, and it's a go 👍

File: bss.cc (hunk 4 above)
Comment (kvoss) [reply bss.cc:565]: umm I think I keep running into some special symbols.. 
But I was mistaken thinking that continuations are necessary in C++. It applies to preprocessor directives, but not to statements, so we are good with this one.

I meant there..
`<SPACE><BACKSLASH>`.

(hopefully that won't get interpreted somehow..)

File: cbench.cc [hunk 13]
```diff
@@ -112,7 +112,7 @@ CBenchMethods[] = {
     {"tripod", py_tripod, METH_VARARGS, "tripod function"},
     {"trefethen4", py_trefethen4, METH_VARARGS, "trefethen4 function"},
     {"three_hump_camel_back", py_three_hump_camel_back, METH_VARARGS, "three hump camel function"},
-    {"dixonPrice", py_dixonPrice, METH_VARARGS, "dixon price function"},
+    {"dixon_price", py_dixonPrice, METH_VARARGS, "dixon price function"},
```
Comment (kvoss) [inline cbench.cc]: and `py_dixon_price`, and that's it 💯

**Review by kvoss — CHANGES_REQUESTED** [review]
one word to fix left 👍

File: cbench.cc (hunk 13 above)
Comment (dillonago) [reply cbench.cc]: Yeah just saw that! My bad.

**Comment by dillonago** [comment]
Completed! Thanks for everything! :)

**Review by kvoss — APPROVED** [review]
lgtm ! Fantastic work! :)

**Pull request merged**